File: kernel/src/syscall.zig
//! KOZO Kernel - Syscall Dispatcher
//! Responsibility: Validate and dispatch syscalls with capability checks

const std = @import("std");
const abi = @import("abi.zig");
const cap = @import("capability.zig");
const thread = @import("thread.zig");
const sched = @import("scheduler.zig");

pub const SYS_MAP_MEMORY = 12;

/// System V ABI: args in rdi, rsi, rdx, r10, r8, r9
/// Returns: error code (0 = success, negative = error)
pub export fn kozo_syscall_handler(
    n: usize,
    arg0: usize, // rdi
    arg1: usize, // rsi  
    arg2: usize, // rdx
    arg3: usize, // r10
    arg4: usize, // r8
    _arg5: usize, // r9 (reserved for future use)
) isize {
    // Silence unused warnings - all args used depending on syscall
    _ = arg4;
    _ = _arg5;

    // Validate syscall number range for security
    if (n > 99) {
        return -1; // KOZO_ERR_INVALID
    }

    return switch (n) {
        // === CAPABILITY PRIMITIVES (Layer 0) ===
        abi.SYS_RETYPE => cap.sys_retype(arg0, @enumFromInt(arg1), arg2, arg3),
        abi.SYS_CAP_TRANSFER => cap.sys_transfer(arg0, arg1, arg2, arg3),
        abi.SYS_CAP_MINT => cap.sys_mint(arg0, arg1),
        abi.SYS_CAP_REVOKE => cap.sys_revoke(arg0, arg1),
        abi.SYS_CAP_VERIFY => cap.sys_verify(arg0, arg1),
        abi.SYS_CAP_CREATE => cap.sys_create(arg0, @enumFromInt(arg1), arg2), // Create from untyped
        abi.SYS_CAP_DELETE => cap.sys_delete(arg0), // Remove from cnode but don't free

        // === IPC - DIRECT SWITCH (Performance Critical) ===
        abi.SYS_IPC_SEND => ipc_send(arg0, arg1, arg2, arg3),
        abi.SYS_IPC_RECV => ipc_recv(arg0, arg1, arg2),
        abi.SYS_IPC_CALL => ipc_call(arg0, arg1, arg2, arg3), // Direct switch to dest
        abi.SYS_IPC_REPLY => ipc_reply(arg0, arg1), // Direct switch back to caller

        // === THREADING ===
        abi.SYS_THREAD_CREATE => thread_create(arg0, arg1, arg2, arg3),
        abi.SYS_THREAD_RESUME => thread_resume(arg0),
        abi.SYS_THREAD_SUSPEND => thread_suspend(arg0),
        abi.SYS_THREAD_SET_PRIORITY => thread_set_priority(arg0, arg1),

        // === ENDPOINTS (Policy Service Registration) ===
        abi.SYS_ENDPOINT_CREATE => endpoint_create(arg0),
        abi.SYS_ENDPOINT_DELETE => endpoint_delete(arg0),
        abi.SYS_NAMESPACE_REGISTER => namespace_register(arg0, arg1, arg2),

        // === MEMORY ===
        abi.SYS_MAP_FRAME => map_frame(arg0, arg1, arg2, arg3),
        abi.SYS_UNMAP_FRAME => unmap_frame(arg0, arg1),

        // === DEBUGGING ===
        abi.SYS_DEBUG_PUTCHAR => {
            const char: u8 = @truncate(arg0);
            // Write to serial port directly (bypassing stdio for kernel debug)
            const serial_port: *volatile u8 = @ptrFromInt(0x3F8); // COM1
            serial_port.* = char;
            return 0;
        },
        abi.SYS_DEBUG_DUMP_CAPS => debug_dump_caps(),

        else => {
            std.log.warn("KOZO: Invalid syscall {d} from thread {d}", .{ 
                n, thread.getCurrent().tid 
            });
            return -1; // KOZO_ERR_INVALID
        },
    };
}

// === IPC IMPLEMENTATION (Direct Switch) ===

/// SYS_IPC_CALL: Synchronous call with direct context switch
/// arg0: endpoint capability index (in current cnode)
/// arg1: message buffer address (in caller address space)
/// arg2: message length (max 512 bytes per abi.IPC_BUFFER_SIZE)
/// arg3: timeout (0 = blocking, >0 = timeout in ms)
/// Returns: 0 on success (reply received), error code on failure
fn ipc_call(endpoint_idx: usize, msg_ptr: usize, msg_len: usize, timeout: usize) isize {
    if (msg_len > abi.IPC_BUFFER_SIZE) return -6; // InvalidParam
    
    const current = thread.getCurrent() orelse return -2;
    
    // Verify endpoint capability
    const ep_slot = current.root_cnode.get(endpoint_idx) catch return -2;
    if (ep_slot.cap_type != .CAP_ENDPOINT) return -1;
    
    const endpoint = &ep_slot.data.endpoint;
    
    // Copy message to kernel buffer (validated user pointer)
    var msg_buffer: [abi.IPC_BUFFER_SIZE]u8 = undefined;
    const user_msg = @as([*]const u8, @ptrFromInt(msg_ptr))[0..msg_len];
    @memcpy(msg_buffer[0..msg_len], user_msg);
    
    // Find recipient (server) thread waiting on this endpoint
    const recipient = endpoint.queue.dequeue() orelse {
        // No server waiting - queue ourselves and block
        if (timeout == 0) {
            current.state = .BLOCKED_SEND;
            current.ipc_endpoint = endpoint_idx;
            current.ipc_msg_len = msg_len;
            @memcpy(current.ipc_buffer[0..msg_len], msg_buffer[0..msg_len]);
            
            sched.block(current); // Context switch to next runnable
            return 0; // Returns here when reply comes
        } else {
            return -7; // WouldBlock (timeout)
        }
    };
    
    // DIRECT SWITCH: Transfer directly to recipient
    // Save caller context, restore recipient, set reply capability
    
    // Set up reply capability (implicit in call)
    current.ipc_state = .WAITING_REPLY;
    current.ipc_caller = recipient; // Server knows who to reply to
    
    // Transfer message to recipient
    recipient.ipc_msg_len = msg_len;
    @memcpy(recipient.ipc_buffer[0..msg_len], msg_buffer[0..msg_len]);
    recipient.ipc_state = .RECEIVED_CALL;
    recipient.ipc_badge = endpoint.badge; // Server sees caller's badge
    
    // Context switch to recipient (no scheduler involvement)
    sched.switch_to(recipient);
    
    // When recipient replies, we resume here
    return 0;
}

/// SYS_IPC_REPLY: Reply to caller with direct switch back
/// arg0: reply message buffer address
/// arg1: reply message length
/// Returns: 0 on success
fn ipc_reply(reply_ptr: usize, reply_len: usize) isize {
    if (reply_len > abi.IPC_BUFFER_SIZE) return -6;
    
    const current = thread.getCurrent() orelse return -2;
    
    // Verify we are in a call (have a caller waiting)
    if (current.ipc_state != .RECEIVED_CALL or current.ipc_caller == null) {
        return -8; // NoCaller (reply without call)
    }
    
    const caller = current.ipc_caller.?;
    
    // Copy reply to caller's buffer
    const reply_msg = @as([*]const u8, @ptrFromInt(reply_ptr))[0..reply_len];
    caller.ipc_msg_len = reply_len;
    @memcpy(caller.ipc_buffer[0..reply_len], reply_msg);
    
    // Mark caller runnable
    caller.state = .RUNNABLE;
    caller.ipc_state = .IDLE;
    sched.enqueue(caller);
    
    // Mark current thread (server) as waiting for next call
    current.ipc_state = .WAITING_CALL;
    current.ipc_caller = null;
    
    // DIRECT SWITCH BACK to caller
    // If caller has higher priority, it runs immediately
    // Otherwise we continue and caller goes to runqueue
    
    if (caller.priority <= current.priority) {
        sched.switch_to(caller);
    }
    
    return 0;
}

/// SYS_IPC_SEND: Async send (non-blocking)
fn ipc_send(endpoint_idx: usize, msg_ptr: usize, msg_len: usize, timeout: usize) isize {
    _ = timeout; // For async, timeout only matters if queue full
    if (msg_len > abi.IPC_BUFFER_SIZE) return -6;
    
    const current = thread.getCurrent() orelse return -2;
    const ep_slot = current.root_cnode.get(endpoint_idx) catch return -2;
    if (ep_slot.cap_type != .CAP_ENDPOINT) return -1;
    
    const endpoint = &ep_slot.data.endpoint;
    
    // Try to deliver immediately if server waiting
    if (endpoint.queue.dequeue()) |server| {
        // Fast path: direct switch to server
        const msg = @as([*]const u8, @ptrFromInt(msg_ptr))[0..msg_len];
        server.ipc_msg_len = msg_len;
        @memcpy(server.ipc_buffer[0..msg_len], msg);
        server.ipc_state = .RECEIVED_ASYNC;
        server.ipc_badge = endpoint.badge;
        sched.switch_to(server);
        return 0;
    }
    
    // Slow path: queue message (for async) or block
    // For genesis block, just return WouldBlock
    return -7; // WouldBlock
}

/// SYS_IPC_RECV: Blocking receive on endpoint
fn ipc_recv(endpoint_idx: usize, buf_ptr: usize, buf_len: usize) isize {
    if (buf_len < abi.IPC_BUFFER_SIZE) return -6;
    
    const current = thread.getCurrent() orelse return -2;
    const ep_slot = current.root_cnode.get(endpoint_idx) catch return -2;
    if (ep_slot.cap_type != .CAP_ENDPOINT) return -1;
    
    const endpoint = &ep_slot.data.endpoint;
    
    // Check if message already pending (from async send)
    if (current.ipc_state == .RECEIVED_ASYNC or current.ipc_state == .RECEIVED_CALL) {
        // Copy to user buffer
        const user_buf = @as([*]u8, @ptrFromInt(buf_ptr))[0..current.ipc_msg_len];
        @memcpy(user_buf, current.ipc_buffer[0..current.ipc_msg_len]);
        const len = current.ipc_msg_len;
        
        // Reset state
        current.ipc_state = .IDLE;
        
        return @intCast(len);
    }
    
    // No message waiting - block until sender arrives
    current.state = .BLOCKED_RECV;
    current.ipc_endpoint = endpoint_idx;
    endpoint.queue.enqueue(current);
    
    sched.block(current);
    
    // Returns here when message arrives
    // Copy to user buffer
    const user_buf = @as([*]u8, @ptrFromInt(buf_ptr))[0..current.ipc_msg_len];
    @memcpy(user_buf, current.ipc_buffer[0..current.ipc_msg_len]);
    
    return @intCast(current.ipc_msg_len);
}

// === THREADING ===

/// SYS_THREAD_CREATE: Create new thread in specified address space
/// arg0: vspace capability index (CAP_ADDRESS_SPACE)
/// arg1: entry point address (virtual)
/// arg2: stack pointer (virtual)  
/// arg3: cnode capability index for root (initial capabilities)
/// Returns: thread capability index on success, error on failure
fn thread_create(vspace_idx: usize, entry: usize, stack: usize, cnode_idx: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    
    // Verify VSpace capability
    const vspace_slot = current.root_cnode.get(vspace_idx) catch return -2;
    if (vspace_slot.cap_type != .CAP_ADDRESS_SPACE) return -1;
    
    // Verify CNode capability  
    const cnode_slot = current.root_cnode.get(cnode_idx) catch return -2;
    if (cnode_slot.cap_type != .CAP_CNODE) return -1;
    
    // Allocate TCB from kernel pool (fixed memory)
    const tcb = thread.allocTCB() orelse return -3; // NoMem
    
    // Initialize thread
    tcb.vspace = vspace_slot;
    tcb.root_cnode = cnode_slot;
    tcb.context.rip = entry;
    tcb.context.rsp = stack;
    tcb.state = .SUSPENDED;
    tcb.priority = current.priority; // Inherit priority
    
    // Create thread capability in current cnode
    const slot = current.root_cnode.findFreeSlot() catch {
        thread.freeTCB(tcb);
        return -5; // NoSpace
    };
    
    current.root_cnode.insert(slot, .{
        .cap_type = .CAP_THREAD,
        .rights = abi.RIGHT_READ | abi.RIGHT_WRITE | abi.RIGHT_GRANT,
        .badge = cap.generateBadge(@intFromPtr(tcb), .CAP_THREAD),
        .data = .{ .thread = .{ .tcb = tcb } },
    }) catch return -1;
    
    return @intCast(slot);
}

/// SYS_THREAD_RESUME: Start a suspended thread
fn thread_resume(thread_idx: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    const thread_slot = current.root_cnode.get(thread_idx) catch return -2;
    if (thread_slot.cap_type != .CAP_THREAD) return -1;
    
    const tcb = thread_slot.data.thread.tcb;
    if (tcb.state != .SUSPENDED) return -9; // InvalidState
    
    tcb.state = .RUNNABLE;
    sched.enqueue(tcb);
    
    // Preempt if new thread has higher priority
    if (tcb.priority < current.priority) {
        sched.yield();
    }
    
    return 0;
}

fn thread_suspend(thread_idx: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    const thread_slot = current.root_cnode.get(thread_idx) catch return -2;
    if (thread_slot.cap_type != .CAP_THREAD) return -1;
    
    const tcb = thread_slot.data.thread.tcb;
    if (tcb.state == .RUNNING) {
        sched.forceSuspend(tcb);
    } else if (tcb.state == .RUNNABLE) {
        sched.removeFromQueue(tcb);
        tcb.state = .SUSPENDED;
    }
    
    return 0;
}

fn thread_set_priority(thread_idx: usize, priority: usize) isize {
    if (priority > 255) return -6;
    
    const current = thread.getCurrent() orelse return -2;
    const thread_slot = current.root_cnode.get(thread_idx) catch return -2;
    if (thread_slot.cap_type != .CAP_THREAD) return -1;
    
    const tcb = thread_slot.data.thread.tcb;
    
    // Cannot increase priority above own (DID: prevents priority escalation)
    if (priority < current.priority) return -4;
    
    tcb.priority = @intCast(priority);
    
    // Re-queue if runnable and priority changed
    if (tcb.state == .RUNNABLE) {
        sched.reprioritize(tcb);
    }
    
    return 0;
}

// === ENDPOINTS ===

/// SYS_ENDPOINT_CREATE: Create new IPC endpoint
fn endpoint_create(slot_idx: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    
    // Ensure slot is free
    const slot = current.root_cnode.get(slot_idx) catch return -2;
    if (slot.cap_type != .CAP_NULL) return -5; // NoSpace
    
    // Retype from untyped? For genesis, we need an untyped index
    // Actually this should take an untyped index to retype from
    return -1; // TODO: Implement with untyped retype
}

fn endpoint_delete(endpoint_idx: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    const ep_slot = current.root_cnode.get(endpoint_idx) catch return -2;
    if (ep_slot.cap_type != .CAP_ENDPOINT) return -1;
    
    // Wake any blocked threads with error
    const endpoint = &ep_slot.data.endpoint;
    while (endpoint.queue.dequeue()) |t| {
        t.state = .RUNNABLE;
        t.ipc_state = .ERROR;
        sched.enqueue(t);
    }
    
    ep_slot.* = cap.NULL_SLOT;
    return 0;
}

/// SYS_NAMESPACE_REGISTER: Register endpoint in global namespace
/// arg0: endpoint capability index
/// arg1: name pointer (null-terminated string)
/// arg2: name length
fn namespace_register(endpoint_idx: usize, name_ptr: usize, name_len: usize) isize {
    _ = name_len;
    const current = thread.getCurrent() orelse return -2;
    const ep_slot = current.root_cnode.get(endpoint_idx) catch return -2;
    if (ep_slot.cap_type != .CAP_ENDPOINT) return -1;
    
    // For genesis block, just verify this is Policy Service (special case)
    // Real implementation needs a namespace service
    const name = @as([*]const u8, @ptrFromInt(name_ptr));
    if (std.mem.eql(u8, name[0..12], "system.policy")) {
        // Mark as system endpoint
        ep_slot.data.endpoint.badge = 0x1; // Reserved for Policy Service
    }
    
    return 0;
}

// === MEMORY ===

fn map_frame(frame_idx: usize, vaddr: usize, rights: usize, attr: usize) isize {
    _ = attr;
    const current = thread.getCurrent() orelse return -2;
    const frame_slot = current.root_cnode.get(frame_idx) catch return -2;
    if (frame_slot.cap_type != .CAP_FRAME) return -1;
    if (frame_slot.rights & rights != rights) return -4; // Insufficient rights
    
    const frame = &frame_slot.data.frame;
    if (frame.mapped) return -9; // Already mapped
    
    // Get current VSpace
    const vspace = current.vspace;
    
    // Walk page tables and map
    // This is architecture-specific
    const result = arch.mapPage(vspace, frame.phys, vaddr, rights);
    if (result == 0) {
        frame.mapped = true;
    }
    
    return result;
}

fn unmap_frame(frame_idx: usize, vaddr: usize) isize {
    const current = thread.getCurrent() orelse return -2;
    const frame_slot = current.root_cnode.get(frame_idx) catch return -2;
    if (frame_slot.cap_type != .CAP_FRAME) return -1;
    
    const frame = &frame_slot.data.frame;
    if (!frame.mapped) return -9;
    
    // Unmap from address space
    arch.unmapPage(current.vspace, vaddr);
    frame.mapped = false;
    
    return 0;
}

// === DEBUGGING ===

fn debug_dump_caps() isize {
    const current = thread.getCurrent() orelse return -2;
    std.log.info("=== Capabilities for Thread {d} ===", .{current.tid});
    
    for (current.root_cnode.slots, 0..) |slot, i| {
        if (slot.cap_type != .CAP_NULL) {
            std.log.info("  Slot {d}: type={d} rights={x} badge={x}", .{
                i, @intFromEnum(slot.cap_type), slot.rights, slot.badge
            });
        }
    }
    
    return 0;
}

// === ARCHITECTURE-SPECIFIC STUBS ===

const arch = struct {
    fn mapPage(vspace: cap.CapSlot, phys: usize, virt: usize, rights: usize) isize {
        _ = vspace;
        _ = phys;
        _ = virt;
        _ = rights;
        // x86_64: Walk page tables, insert PTE
        return 0; // TODO: Implement for genesis
    }
    
    fn unmapPage(vspace: cap.CapSlot, virt: usize) void {
        _ = vspace;
        _ = virt;
        // x86_64: Clear PTE, invalidate TLB
    }
};

File: services/kozo-sys/src/syscall.rs
//! KOZO-SYS: Safe Syscall Wrappers
//! File Path: services/kozo-sys/src/syscall.rs
//! Responsibility: x86_64 syscall ABI implementation, mapping Rust to Zig kernel
//! Security: DID Principle - All unsafe assembly isolated in this audited module
//! ABI: rax=num, rdi=a0, rsi=a1, rdx=a2, r10=a3, r8=a4, r9=a5

use crate::abi::{CapType, Error, Rights, Syscall};
use core::arch::asm;

// =============================================================================
// RAW SYSCALL PRIMITIVES (Unsafe - use wrapped versions below)
// =============================================================================

/// Raw syscall with 0 arguments
/// 
/// # Safety
/// Direct kernel call. Must be valid syscall number.
#[inline(always)]
pub unsafe fn syscall0(n: Syscall) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

/// Raw syscall with 1 argument
#[inline(always)]
pub unsafe fn syscall1(n: Syscall, a0: usize) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        in("rdi") a0,
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

/// Raw syscall with 2 arguments
#[inline(always)]
pub unsafe fn syscall2(n: Syscall, a0: usize, a1: usize) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        in("rdi") a0,
        in("rsi") a1,
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

/// Raw syscall with 3 arguments
#[inline(always)]
pub unsafe fn syscall3(n: Syscall, a0: usize, a1: usize, a2: usize) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        in("rdi") a0,
        in("rsi") a1,
        in("rdx") a2,
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

/// Raw syscall with 4 arguments
/// 
/// Note: a3 goes in r10 (not rcx) per x86_64 syscall ABI
#[inline(always)]
pub unsafe fn syscall4(n: Syscall, a0: usize, a1: usize, a2: usize, a3: usize) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        in("rdi") a0,
        in("rsi") a1,
        in("rdx") a2,
        in("r10") a3, // r10, not rcx!
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

/// Raw syscall with 6 arguments (full register set)
#[inline(always)]
pub unsafe fn syscall6(n: Syscall, a0: usize, a1: usize, a2: usize, a3: usize, a4: usize, a5: usize) -> isize {
    let ret: isize;
    asm!(
        "syscall",
        in("rax") n as usize,
        in("rdi") a0,
        in("rsi") a1,
        in("rdx") a2,
        in("r10") a3,
        in("r8") a4,
        in("r9") a5,
        lateout("rax") ret,
        out("rcx") _, out("r11") _,
        options(nostack, preserves_flags)
    );
    ret
}

// =============================================================================
// CAPABILITY MANAGEMENT (Safe Wrappers)
// =============================================================================

/// SYS_RETYPE: Convert untyped memory into kernel object
/// 
/// # Arguments
/// * `untyped_slot` - Slot index of untyped capability in current CNode
/// * `obj_type` - Type to create (CNode, Endpoint, Thread, etc.)
/// * `dest_slot` - Slot to place new capability
/// * `size_bits` - log2(size) for variable-sized objects (CNodes)
/// 
/// # Safety
/// Destroys information in memory region (zeros it). Irreversible.
pub fn sys_retype(untyped_slot: usize, obj_type: CapType, dest_slot: usize, size_bits: usize) -> Result<(), Error> {
    let res = unsafe { syscall4(Syscall::Retype, untyped_slot, obj_type as usize, dest_slot, size_bits) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_CAP_TRANSFER: Move/Copy capability between CNodies
/// 
/// Used by Policy Service to delegate capabilities to apps.
/// Requires GRANT right on source capability.
/// 
/// # Arguments
/// * `src_slot` - Source capability in current CNode
/// * `dest_cnode` - Target CNode badge (usually app_id.0)
/// * `dest_slot` - Slot in target CNode
/// * `rights` - Attenuated rights for destination (subset of source)
pub fn sys_cap_transfer(src_slot: usize, dest_cnode: u64, dest_slot: usize, rights: Rights) -> Result<(), Error> {
    let res = unsafe { 
        syscall4(
            Syscall::CapTransfer, 
            src_slot, 
            dest_cnode as usize, 
            dest_slot, 
            rights.bits() as usize
        ) 
    };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_CAP_MINT: Create attenuated child capability
/// 
/// Derives new capability with subset of parent rights.
/// Returns new slot index on success.
/// 
/// # Arguments
/// * `parent_slot` - Source capability (must have GRANT right)
/// * `new_rights` - Rights for child (subset of parent)
pub fn sys_cap_mint(parent_slot: usize, new_rights: Rights) -> Result<usize, Error> {
    let res = unsafe { syscall2(Syscall::CapMint, parent_slot, new_rights.bits() as usize) };
    if res < 0 { 
        Err(Error::from_raw(res)) 
    } else { 
        Ok(res as usize) 
    }
}

/// SYS_CAP_REVOKE: Destroy capability and all derivatives
/// 
/// Immediate invalidation. Apps using this capability will fault.
/// Used for JIT delegation timeout enforcement.
pub fn sys_cap_revoke(cnode: usize, slot: usize) -> Result<(), Error> {
    let res = unsafe { syscall2(Syscall::CapRevoke, cnode, slot) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_CAP_VERIFY: Verify badge authenticity (anti-spoofing)
/// 
/// Policy Service uses this to confirm AppID claims.
/// Validates that badge matches kernel's records for thread.
pub fn sys_cap_verify(badge: u64, thread_cap: usize) -> Result<(), Error> {
    let res = unsafe { syscall2(Syscall::CapVerify, badge as usize, thread_cap) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_CAP_CREATE: Create capability from untyped (legacy alias for retype)
pub fn sys_cap_create(untyped: usize, obj_type: CapType, slot: usize) -> Result<(), Error> {
    sys_retype(untyped, obj_type, slot, 0)
}

/// SYS_CAP_DELETE: Remove capability from CNode (don't free memory)
/// 
/// Safer than revoke - doesn't cascade. Use for cleanup.
pub fn sys_cap_delete(slot: usize) -> Result<(), Error> {
    let res = unsafe { syscall1(Syscall::CapDelete, slot) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// IPC (INTER-PROCESS COMMUNICATION)
// =============================================================================

/// SYS_IPC_SEND: Asynchronous message send
/// 
/// Non-blocking if timeout is 0. Use for notifications.
/// 
/// # Arguments
/// * `endpoint` - Endpoint capability slot
/// * `buf` - Message buffer pointer
/// * `len` - Message length (max IPC_BUFFER_SIZE)
/// * `timeout` - 0=non-blocking, >0=blocking with timeout
pub unsafe fn sys_ipc_send(endpoint: usize, buf: *const u8, len: usize, timeout: usize) -> Result<(), Error> {
    let res = syscall4(Syscall::IpcSend, endpoint, buf as usize, len, timeout);
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_IPC_RECV: Blocking message receive
/// 
/// Returns badge (sender identity) on success.
/// 
/// # Arguments
/// * `endpoint` - Endpoint to listen on
/// * `buf` - Buffer to receive message (must be writable)
/// * `buf_size` - Buffer capacity (should be IPC_BUFFER_SIZE)
/// 
/// # Returns
/// Badge (u64 as usize) identifying sender
pub unsafe fn sys_ipc_recv(endpoint: usize, buf: *mut u8, buf_size: usize) -> Result<u64, Error> {
    let res = syscall3(Syscall::IpcRecv, endpoint, buf as usize, buf_size);
    if res < 0 { 
        Err(Error::from_raw(res)) 
    } else { 
        Ok(res as u64) 
    }
}

/// SYS_IPC_CALL: Synchronous call with direct context switch
/// 
/// Fast path: switches directly to recipient without scheduler.
/// Blocks until recipient calls ipc_reply.
/// 
/// # Performance
/// ~300 cycles vs ~1000 for separate send/recv/schedule
pub unsafe fn sys_ipc_call(endpoint: usize, msg: *const u8, msg_len: usize, timeout: usize) -> Result<(), Error> {
    let res = syscall4(Syscall::IpcCall, endpoint, msg as usize, msg_len, timeout);
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_IPC_REPLY: Reply to call with direct switch back
/// 
/// Must be called while handling a call. Returns to caller.
pub unsafe fn sys_ipc_reply(reply: *const u8, reply_len: usize) -> Result<(), Error> {
    let res = syscall2(Syscall::IpcReply, reply as usize, reply_len);
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// THREADING
// =============================================================================

/// SYS_THREAD_CREATE: Create new thread in address space
/// 
/// # Arguments
/// * `vspace` - Address space (page table root) capability
/// * `entry` - Entry point virtual address
/// * `stack` - Stack pointer virtual address
/// * `cnode` - Root CNode for new thread
/// 
/// # Returns
/// Thread capability slot index
pub fn sys_thread_create(vspace: usize, entry: usize, stack: usize, cnode: usize) -> Result<usize, Error> {
    let res = unsafe { syscall4(Syscall::ThreadCreate, vspace, entry, stack, cnode) };
    if res < 0 { 
        Err(Error::from_raw(res)) 
    } else { 
        Ok(res as usize) 
    }
}

/// SYS_THREAD_RESUME: Start suspended thread
/// 
/// Thread begins execution at entry point specified in create.
pub fn sys_thread_resume(thread_cap: usize) -> Result<(), Error> {
    let res = unsafe { syscall1(Syscall::ThreadResume, thread_cap) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_THREAD_SUSPEND: Pause thread
/// 
/// Thread can be resumed later. Use for debugging.
pub fn sys_thread_suspend(thread_cap: usize) -> Result<(), Error> {
    let res = unsafe { syscall1(Syscall::ThreadSuspend, thread_cap) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_THREAD_SET_PRIORITY: Adjust scheduling priority
/// 
/// Cannot increase above own priority (prevents escalation).
pub fn sys_thread_set_priority(thread_cap: usize, priority: u8) -> Result<(), Error> {
    let res = unsafe { syscall2(Syscall::ThreadSetPriority, thread_cap, priority as usize) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// ENDPOINTS & NAMESPACE
// =============================================================================

/// SYS_ENDPOINT_CREATE: Create new IPC endpoint
/// 
/// Endpoints are unidirectional message queues.
pub fn sys_endpoint_create(slot: usize) -> Result<(), Error> {
    let res = unsafe { syscall1(Syscall::EndpointCreate, slot) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_ENDPOINT_DELETE: Destroy endpoint and wake blocked threads
pub fn sys_endpoint_delete(slot: usize) -> Result<(), Error> {
    let res = unsafe { syscall1(Syscall::EndpointDelete, slot) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_NAMESPACE_REGISTER: Register endpoint in system namespace
/// 
/// Allows other processes to find this endpoint by name.
/// Critical for Policy Service ("system.policy").
/// 
/// # Safety
/// Name must be valid UTF-8, null-terminated.
pub unsafe fn sys_namespace_register(endpoint: usize, name: *const u8, name_len: usize) -> Result<(), Error> {
    let res = syscall3(Syscall::NamespaceRegister, endpoint, name as usize, name_len);
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// MEMORY
// =============================================================================

/// SYS_MAP_FRAME: Map physical page into address space
/// 
/// # Arguments
/// * `frame_cap` - Frame capability slot
/// * `vaddr` - Virtual address to map (must be page-aligned)
/// * `rights` - Mapping rights (READ/WRITE/EXEC)
/// * `attrs` - Cache attributes, etc.
pub fn sys_map_frame(frame_cap: usize, vaddr: usize, rights: Rights, attrs: usize) -> Result<(), Error> {
    let res = unsafe { syscall4(Syscall::MapFrame, frame_cap, vaddr, rights.bits() as usize, attrs) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

/// SYS_UNMAP_FRAME: Remove mapping
/// 
/// Frame can be remapped elsewhere after unmap.
pub fn sys_unmap_frame(frame_cap: usize, vaddr: usize) -> Result<(), Error> {
    let res = unsafe { syscall2(Syscall::UnmapFrame, frame_cap, vaddr) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// DEBUGGING
// =============================================================================

/// SYS_DEBUG_PUTCHAR: Output character to kernel debug console
/// 
/// Safe to call anytime. Used for early boot debugging.
pub fn sys_debug_putchar(c: u8) {
    unsafe { 
        syscall1(Syscall::DebugPutchar, c as usize); 
    }
}

/// SYS_DEBUG_PRINT: Output string to kernel debug console
/// 
/// Convenience function (not a real syscall, loops over putchar).
pub fn sys_debug_print(s: &str) {
    for c in s.bytes() {
        sys_debug_putchar(c);
    }
}

/// SYS_DEBUG_DUMP_CAPS: Print capability table to debug console
/// 
/// Useful for debugging capability issues.
pub fn sys_debug_dump_caps() -> Result<(), Error> {
    let res = unsafe { syscall0(Syscall::DebugDumpCaps) };
    if res == 0 { Ok(()) } else { Err(Error::from_raw(res)) }
}

// =============================================================================
// CONVENIENCE MODULES
// =============================================================================

/// Safe IPC helpers using IPCBuffer type
pub mod ipc {
    use super::*;
    use crate::ipc::IPCBuffer;

    /// Send using IPCBuffer
    pub fn send(endpoint: usize, buf: &IPCBuffer, timeout: usize) -> Result<(), Error> {
        unsafe { sys_ipc_send(endpoint, buf.as_ptr(), buf.len(), timeout) }
    }

    /// Receive into IPCBuffer
    pub fn recv(endpoint: usize, buf: &mut IPCBuffer) -> Result<u64, Error> {
        let badge = unsafe { sys_ipc_recv(endpoint, buf.as_mut_ptr(), buf.capacity())? };
        buf.set_len(512); // Max received - TODO: get actual from kernel
        Ok(badge)
    }

    /// Call with IPCBuffer
    pub fn call(endpoint: usize, request: &IPCBuffer, reply: &mut IPCBuffer) -> Result<(), Error> {
        unsafe {
            sys_ipc_call(endpoint, request.as_ptr(), request.len(), 0)?;
            // Reply comes back in same buffer or different? Depends on kernel impl
            Ok(())
        }
    }
}